packages/react/src/menu/root/useMenuRoot.ts

```diff
@@ -134,22 +159,47 @@ export function useMenuRoot(parameters: useMenuRoot.Parameters): useMenuRoot.Ret
     return clearStickIfOpenTimeout;
   }, [clearStickIfOpenTimeout]);
 
-  const floatingRootContext = useFloatingRootContext({
-    elements: {
-      reference: triggerElement,
-      floating: positionerElement,
-    },
-    open,
-    onOpenChange(openValue, eventValue, reasonValue) {
-      const isHover = reasonValue === 'hover' || reasonValue === 'safe-polygon';
-      const isKeyboardClick = reasonValue === 'click' && (eventValue as MouseEvent).detail === 0;
-      const isDismissClose = !openValue && (reasonValue === 'escape-key' || reasonValue == null);
+  const ignoreClickRef = React.useRef(false);
+
+  const setOpen = useEventCallback(
+    (nextOpen: boolean, event: Event | undefined, reason: OpenChangeReason | undefined) => {
+      if (open === nextOpen) {
+        return;
+      }
+
+      // As the menu opens on mousedown and closes on click,
+      // we need to ignore the click event immediately following mousedown.
+      if (event?.type === 'click' && ignoreClickRef.current) {
+        ignoreClickRef.current = false;
+        return;
+      }
+
+      if (nextOpen && event?.type === 'mousedown') {
+        ignoreClickRef.current = true;
+
+        ownerDocument(event.currentTarget as Element).addEventListener(
+          'click',
+          () => {
+            ignoreClickRef.current = false;
+          },
+          { once: true },
+        );
+      } else {
+        ignoreClickRef.current = false;
+      }
+
+      const isKeyboardClick =
+        (reason === 'click' || reason === 'item-press') && (event as MouseEvent).detail === 0;
+      const isDismissClose = !nextOpen && (reason === 'escape-key' || reason == null);
 
       function changeState() {
-        setOpen(openValue, eventValue, translateOpenChangeReason(reasonValue));
+        onOpenChange?.(nextOpen, event, reason);
+        setOpenUnwrapped(nextOpen);
+
+        setLastOpenChangeReason(reason ?? null);
       }
 
-      if (isHover) {
+      if (reason === 'hover') {
```

Comment:
Hover (leave) events can also generate `'safe-polygon'`

Reply:
We're translating Floating UI's `safe-polygon` to `hover`, so this should be covered.